qiskit_experiments/library/characterization/resonator_spectroscopy.py

```diff
@@ -0,0 +1,218 @@
+
+        Raises:
+            QiskitError: if no frequencies are given and absolute frequencies are desired and
+                no backend is given.
+        """
+        analysis = ResonatorSpectroscopyAnalysis()
+
+        if frequencies is None:
+            frequencies = np.linspace(-20.0e6, 20.0e6, 51)
+
+            if absolute:
+                if backend is None:
+                    raise QiskitError(
+                        "Cannot automatically compute absolute frequencies without a backend."
+                    )
+
+                center_freq = backend.defaults().meas_freq_est[qubit]
+                frequencies += center_freq
+
+        super().__init__(qubit, frequencies, backend, absolute, analysis, **experiment_options)
+
+    @property
+    def _backend_center_frequency(self) -> float:
+        """Returns the center frequency of the experiment.
+
+        Returns:
+            The center frequency of the experiment.
+
+        Raises:
+            QiskitError: If the experiment does not have a backend set.
+        """
+        if self.backend is None:
+            raise QiskitError("backend not set. Cannot call center_frequency.")
+
+        return self.backend.defaults().meas_freq_est[self.physical_qubits[0]]
+
+    def _template_circuit(self) -> QuantumCircuit:
+        """Return the template quantum circuit."""
+        circuit = QuantumCircuit(1, 1)
+        circuit.measure(0, 0)
+
+        return circuit
+
+    def _schedule(self) -> Tuple[pulse.ScheduleBlock, Parameter]:
+        """Create the spectroscopy schedule."""
+
+        dt = getattr(self.backend.configuration(), "dt", None)
+        constraints = getattr(self.backend.configuration(), "timing_constraints", {})
+        granularity = constraints.get("granularity", None)
+
+        if dt is None or granularity is None:
+            raise QiskitError(f"{self.__class__.__name__} needs both dt and sample granularity.")
+
+        acq_dur = int(granularity * (self.experiment_options.acquire_duration / dt // granularity))
+        acq_del = int(granularity * (self.experiment_options.acquire_delay / dt // granularity))
+        duration = int(granularity * (self.experiment_options.duration / dt // granularity))
+        sigma = granularity * (self.experiment_options.sigma / dt // granularity)
+        width = granularity * (self.experiment_options.width / dt // granularity)
+
+        qubit = self.physical_qubits[0]
+
+        freq_param = Parameter("frequency")
+
+        with pulse.build(backend=self.backend, name="spectroscopy") as schedule:
+            pulse.shift_frequency(freq_param, pulse.MeasureChannel(qubit))
+            pulse.play(
+                pulse.GaussianSquare(
+                    duration=duration,
+                    amp=self.experiment_options.amp,
+                    sigma=sigma,
+                    width=width,
+                ),
+                pulse.MeasureChannel(qubit),
+            )
+
+            with pulse.align_left():
```

Comment:
I had this comment before but it was outdated by a code refactor:



This is the sensible way to parametrize this. We should test on live backends. I believe that at least for IBM Quantum backends with older hardware you can not begin an acquisition in the middle of a measurement pulse (you can split the pulse into two). Also, the acquisition duration currently does not do anything. We might consider not allowing the delay or duration parameters for now to avoid users dealing with odd backend limitations.

To some degree these issues can be worked around using a custom measurement kernel but that has its own difficulties because the API for setting a custom kernel is not easy to work with.

Reply:
One way forward would be to remove the delay option for the acquire channel in this PR. Then, when backends are ready for this we can reintroduce the option. It's only a small change. How does that sound?

Reply:
Yes, that seems fine to me.

Reply:
Done here: 9358a2b